clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```diff
@@ -243,6 +244,11 @@
      */
     Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions, Duration timeout);
 
+    /**
+     * @see KafkaConsumer#currentLag(TopicPartition)
+     */
+    OptionalLong currentLag(TopicPartition topicPartition);
```

Comment:
Pardon me, KIP-695 does not include this change. It seems KIP-695 is still based on `metadata`? Please correct me If I misunderstand anything :)

Reply:
Woah, you are _fast_, @chia7712 !

I just sent a message to the vote thread. I wanted to submit this PR first so that the vote thread message can have the full context available.

Do you mind reading over what I said there? If it sounds good to you, then I'll update the KIP, and we can maybe put this whole mess to bed.

Reply:
Quick question, should the API take a `Collection<TopicPartition>` like other APIs?

Reply:
Thanks @ijuma, I considered it, but decided on the current API because:
1. This is a very quick, local in-memory lookup, so there's no reason to batch multiple requests in one
2. It complicates the return type. We'd have to return either a `Map<TP, Long>`, with mappings missing for unknown lags (which creates unfortunate null semantics for users), or a `Map<TP, OptionalLong>` which creates a complex-to-understand two hop lookup (`lag:=result.get(tp).get()`). Or else, we could return a more complex domain object object like @chia7712 proposed in the mailing list. All these complicates seem like unnecessary complexity in the case of this particular API, given the first point.

WDYT?

Reply:
If there is no batching benefit, then the simpler API makes sense. @hachikuji Any reason why batching could be useful here?

Reply:
For API calls that may incur a broker round trip, have batching of partitions makes sense. For this API I think single partition lookup is good enough.

Reply:
If we concern that users may call this function too frequent looping a large number of partitions, and each call is synchronizing on the subscription state, then maybe we can make it in a batching mode.

Reply:
That's a good point, @guozhangwang. Entering the synchronized block will have some overhead each time it's called.

I think we can just reason about the use cases here. My guess is that people would either tend to spot-check specific lags, as we are doing here, or they would tend to periodically check all lags. In the former case, I'd hazard that the current API is fine. In the latter case, we'd face more overhead. I'm sure this is motivated reasoning, but perhaps we can lump the latter case in with @chia7712 's suggestion to expose more metadata and defer it to the future.